lib/scholar/spatial/distances.ex

```diff
@@ -0,0 +1,271 @@
+defmodule Scholar.Spatial.Distances do
+  @moduledoc """
+  Distance metrics between 1-D tensors.
+  """
+
+  import Nx.Defn
+  import Scholar.Shared
+
+  @doc """
+  Standard euclidean distance.
+
+  $$
+  D(x, y) = \\sqrt{\\sum_i (x_i - y_i)^2}
+  $$
+
+  ## Examples
+
+      iex> x = Nx.tensor([1, 2])
+      iex> y = Nx.tensor([3, 2])
+      iex> Scholar.Spatial.Distances.euclidean(x, y)
+      #Nx.Tensor<
+        f32
+        2.0
+      >
+
+      iex> x = Nx.tensor([1, 2])
+      iex> y = Nx.tensor([1, 2])
+      iex> Scholar.Spatial.Distances.euclidean(x, y)
+      #Nx.Tensor<
+        f32
+        0.0
+      >
+
+      iex> x = Nx.tensor([1, 2])
+      iex> y = Nx.tensor([1, 2, 3])
+      iex> Scholar.Spatial.Distances.euclidean(x, y)
+      ** (ArgumentError) expected input shapes to be equal, got {2} != {3}
+  """
+  @spec euclidean(Nx.t(), Nx.t()) :: Nx.t()
+  defn euclidean(x, y) do
+    assert_shape!(x, y)
+    diff = x - y
+
+    cond do
+      Nx.all(diff == 0) ->
+        Nx.tensor(0.0)
+
+      true ->
+        diff
+        |> Nx.LinAlg.norm()
+        |> Nx.as_type({:f, 32})
+    end
+  end
+
+  @doc """
+  Squared euclidean distance.
+
+  $$
+  D(x, y) = \\sum_i (x_i - y_i)^2
+  $$
+
+  ## Examples
+
+      iex> x = Nx.tensor([1, 2])
+      iex> y = Nx.tensor([3, 2])
+      iex> Scholar.Spatial.Distances.squared_euclidean(x, y)
+      #Nx.Tensor<
+        f32
+        4.0
+      >
+
+      iex> x = Nx.tensor([1, 2])
+      iex> y = Nx.tensor([1, 2])
+      iex> Scholar.Spatial.Distances.squared_euclidean(x, y)
+      #Nx.Tensor<
+        f32
+        0.0
+      >
+
+      iex> x = Nx.tensor([1, 2])
+      iex> y = Nx.tensor([1, 2, 3])
+      iex> Scholar.Spatial.Distances.squared_euclidean(x, y)
+      ** (ArgumentError) expected input shapes to be equal, got {2} != {3}
+  """
+  @spec squared_euclidean(Nx.t(), Nx.t()) :: Nx.t()
+  defn squared_euclidean(x, y) do
+    assert_shape!(x, y)
+
+    x
+    |> Nx.subtract(y)
+    |> Nx.power(2)
+    |> Nx.sum()
+    |> Nx.as_type({:f, 32})
+  end
+
+  @doc """
+  Manhattan, taxicab, or l1 distance.
+
+  $$
+  D(x, y) = \\sum_i |x_i - y_i|
+  $$
+
+  ## Examples
+
+      iex> x = Nx.tensor([1, 2])
+      iex> y = Nx.tensor([3, 2])
+      iex> Scholar.Spatial.Distances.manhattan(x, y)
+      #Nx.Tensor<
+        f32
+        2.0
+      >
+
+      iex> x = Nx.tensor([1, 2])
+      iex> y = Nx.tensor([1, 2])
+      iex> Scholar.Spatial.Distances.manhattan(x, y)
+      #Nx.Tensor<
+        f32
+        0.0
+      >
+
+      iex> x = Nx.tensor([1, 2])
+      iex> y = Nx.tensor([1, 2, 3])
+      iex> Scholar.Spatial.Distances.manhattan(x, y)
+      ** (ArgumentError) expected input shapes to be equal, got {2} != {3}
+  """
+  @spec manhattan(Nx.t(), Nx.t()) :: Nx.t()
+  defn manhattan(x, y) do
+    assert_shape!(x, y)
+
+    x
+    |> Nx.subtract(y)
+    |> Nx.abs()
```

Comment:
We have Nx.LinAlg.norm(x, 1) or something like that

Reply:
I actually don't think this is a vector norm?

Reply:
It's the L1 norm

Reply:
> I actually don't think this is a vector norm?

Could you elaborate on this?

Reply:
Sorry I got confused by where the review comment was. :facepalm:

Reply:
Replacing `Nx.abs()` and `Nx.sum()` with `Nx.LinAlg.norm(ord: 1)` is throwing for me though when `x == y`. I'm guessing it's a divide by zero error but opaque:

```elixir
     stacktrace:
       (nx 0.1.0) lib/nx/binary_backend.ex:654: Nx.BinaryBackend.element_divide/3
       (nx 0.1.0) lib/nx/binary_backend.ex:612: Nx.BinaryBackend."-element_wise_bin_op/4-lbc$^17/2-9-"/7
       (nx 0.1.0) lib/nx/binary_backend.ex:611: Nx.BinaryBackend.element_wise_bin_op/4
       (nx 0.1.0) lib/nx/defn/evaluator.ex:149: Nx.Defn.Evaluator.eval_apply/4
       (nx 0.1.0) lib/nx/defn/evaluator.ex:63: Nx.Defn.Evaluator.eval/3
       (elixir 1.13.3) lib/enum.ex:1715: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
       (nx 0.1.0) lib/nx/defn/evaluator.ex:130: Nx.Defn.Evaluator.eval_apply/4
       (nx 0.1.0) lib/nx/defn/evaluator.ex:63: Nx.Defn.Evaluator.eval/3
       (elixir 1.13.3) lib/enum.ex:1715: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
       (nx 0.1.0) lib/nx/defn/evaluator.ex:130: Nx.Defn.Evaluator.eval_apply/4
       (nx 0.1.0) lib/nx/defn/evaluator.ex:63: Nx.Defn.Evaluator.eval/3
       (nx 0.1.0) lib/nx/defn/evaluator.ex:30: Nx.Defn.Evaluator.__jit__/4
       (for doctest at) lib/scholar/spatial/distances.ex:115: (test)
```